// canvas-utils/es5-generated/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
       value: true
});
exports.createCanvasEventEmitter = exports.fromImageData = exports.fromContext = exports.fromCanvas = exports.fromImage = exports.createCanvas = exports.canvasToImage = exports.convertEventCoords = exports.rotateContextAt = exports.degreesToRadians = exports.radiansToDegrees = undefined;

var _canvasUtils = require('./lib/canvas-utils');

Object.defineProperty(exports, 'radiansToDegrees', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.radiansToDegrees;
       }
});
Object.defineProperty(exports, 'degreesToRadians', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.degreesToRadians;
       }
});
Object.defineProperty(exports, 'rotateContextAt', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.rotateContextAt;
       }
});
Object.defineProperty(exports, 'convertEventCoords', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.convertEventCoords;
       }
});
Object.defineProperty(exports, 'canvasToImage', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.canvasToImage;
       }
});
Object.defineProperty(exports, 'createCanvas', {
       enumerable: true,
       get: function get() {
              return _canvasUtils.createCanvas;
       }
});

var _imageData = require('./lib/imageData');

Object.defineProperty(exports, 'fromImage', {
       enumerable: true,
       get: function get() {
              return _imageData.fromImage;
       }
});
Object.defineProperty(exports, 'fromCanvas', {
       enumerable: true,
       get: function get() {
              return _imageData.fromCanvas;
       }
});
Object.defineProperty(exports, 'fromContext', {
       enumerable: true,
       get: function get() {
              return _imageData.fromContext;
       }
});
Object.defineProperty(exports, 'fromImageData', {
       enumerable: true,
       get: function get() {
              return _imageData.fromImageData;
       }
});

var _canvasEventEmitter = require('./lib/canvasEventEmitter');

var _canvasEventEmitter2 = _interopRequireDefault(_canvasEventEmitter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.createCanvasEventEmitter = _canvasEventEmitter2.default;
